lambda-buffers-compiler/src/LambdaBuffers/Compiler/TypeClass/Pat.hs

```diff
@@ -55,33 +64,106 @@ data Pat
   | AppP Pat Pat {- Pattern for Type applications -}
   | {- This last one is a bit special. This represents a complete type declaration.
        The first Pat should be instantiated to `Name l` where l is a concrete name.
-       The second Pat should be instantiated to a Pat-List (using :*/Nil) which only contains Names.
+       The second Pat should be instantiated to a Pat-List (using :*/Nil) which only contains TyVarPs.
        The final Pat should be instantiated to a Pat body.
        In some languages, parts of this may be ignored. E.g. in Rust the type name doesn't matter (we use the constr name of the
        outermost inner sum for constructing types). -}
     DecP Pat Pat Pat
   deriving stock (Show, Eq, Ord)
 
-infixr 5 :*
+-- infixr 5 :*
 
 {- Utility functions. Turn a list of types into a product/record/sum type.
 -}
-toProd :: [Pat] -> Pat
-toProd = ProdP . foldr (:*) Nil
+toProdP :: [Pat] -> Pat
+toProdP = ProdP . foldr ConsP NilP
 
-toRec :: [Pat] -> Pat
-toRec = RecP . foldr (:*) Nil
+toRecP :: [Pat] -> Pat
+toRecP = RecP . foldr ConsP NilP
 
-toSum :: [Pat] -> Pat
-toSum = SumP . foldr (:*) Nil
+toSumP :: [Pat] -> Pat
+toSumP = SumP . foldr ConsP NilP
 
 {- Converts a pattern that consists of a well formed pattern list
    (i.e. patterns formed from :* and Nil) into a list of patterns.
 -}
 patList :: Pat -> Maybe [Pat]
 patList = \case
-  Nil -> Just []
-  p1 :* p2 -> (p1 :) <$> patList p2
+  NilP -> Just []
+  p1 `ConsP` p2 -> (p1 :) <$> patList p2
+  _ -> Nothing
+
+data Exp
+  = LitE Literal
+  | NilE
+  | ConsE Exp Exp
+  | LabelE Exp Exp
+  | RecE Exp
+  | ProdE Exp
+  | SumE Exp
+  | -- NO EXPRESSION VARS! EXPRESSIONS DON'T HAVE HOLES!
+    RefE Exp Exp {- 1st arg should be a ModuleName  -}
+  | AppE Exp Exp {- Pattern for Type applications -}
+  | DecE Exp Exp Exp
+  deriving stock (Show, Eq, Ord)
+
+class ExpressionLike (p :: Type) where
+  (*:) :: p -> p -> p
+
+  nil :: p
+
+  (*=) :: p -> p -> p
+
+infixr 5 *:
+
+instance ExpressionLike Pat where
+  p1 *: p2 = p1 `ConsP` p2
+
+  nil = NilP
+
+  p1 *= p2 = LabelP p1 p2
+
+instance ExpressionLike Exp where
+  p1 *: p2 = p1 `ConsE` p2
+
+  nil = NilE
+
+  p1 *= p2 = LabelE p1 p2
+
+{- Utility functions. Turn a list of types into a product/record/sum type.
+-}
+toProdE :: [Exp] -> Exp
+toProdE = ProdE . foldr ConsE NilE
+
+toRecE :: [Exp] -> Exp
+toRecE = RecE . foldr ConsE NilE
+
+toSumE :: [Exp] -> Exp
+toSumE = SumE . foldr ConsE NilE
+
+-- | Extract the "type function" from an AppE Expr
+tyFunE :: Exp -> Maybe Exp
+tyFunE = \case
+  AppE e1 _ -> case tyFunE e1 of
+    Nothing -> Just e1
+    Just e2 -> Just e2
+  _ -> Nothing
```

Comment:
It fails if you pass it an `Exp` that's not a `TyFun` and it's recursive because the point of this is to extract the "initial" (don't know how to express that better) TyFun in an application. Like it gets the `f` in 

```
(f `AppE` x) `AppE` y
```